src/interp/interp.cc

```diff
@@ -934,24 +938,33 @@ void Instance::Mark(Store& store) {
 }
 
 //// Thread ////
-Thread::Thread(Store& store, const Options& options)
-    : Object(skind), store_(store) {
+Thread::Thread(Store& store, Stream* trace_stream)
+    : store_(store), trace_stream_(trace_stream) {
+  prev_thread_ = store.threads_;
+  store.threads_ = this;
+
+  Thread::Options options;
   frames_.reserve(options.call_stack_size);
   values_.reserve(options.value_stack_size);
   trace_stream_ = options.trace_stream;
-  if (options.trace_stream) {
+  if (trace_stream) {
     trace_source_ = MakeUnique<TraceSource>(this);
   }
 }
 
-void Thread::Mark(Store& store) {
+Thread::~Thread()
+{
+  store_.threads_ = prev_thread_;
```

Comment:
Is it not possible for threads to be destroyed in a different order to that in which they were created?

Reply:
It seems to me Thread objects are created on wasm function call, and so the compiler ensures they are destroyed in a reversed creation order by the compiler (unless longjmp is used, but that is discouraged to use in C++). I think the compiler would complain if a Thread is created in another way since the constructor is changed.

If an ECMAScript async like feature would be implemented in wasm in the future, that would likely need a context object, and Thread can be a member of that object. These Threads would be marked through the context object, and would not be part of this Thread stack chain.

Reply:
I guess I'm not sure what this Thread object is then.. surely a thread (at least in the normal sense) is not created for every call but contains an entire calls stack of calls?    And there is no normally a stack relationship between running threads (I would imagine threads is more of a parent child relationship).

Reply:
Oh I see .. there are two ways to `Call` a function:

```
  Result Call(Thread& thread,                                                       
              const Values& params,                                                 
              Values& results,                                                      
              Trap::Ptr* out_trap);                                                 
                                                                                    
  // Convenience function that creates new Thread.                                  
  Result Call(Store&,                                                                                                                                                
              const Values& params,                                                 
              Values& results,                                                   
              Trap::Ptr* out_trap,                                               
              Stream* = nullptr);    
```

The second one create a brand new thread... which I guess should be destroyed when the call returns.

The first one is for when a normal function call is make within an existing thread.

Reply:
I this case I don't think we need to model threads in some kind of stack since their lifetime is independent of of each other, no?

Reply:
Honestly, I don't know the concept behind the design. The `Result Call(Thread& thread, ...)` variant is never called directly, probably users should not create threads directly. Maybe it is a future-proof design for something (my guess is async). But this code improves the memory management of the common use case, and we can think something for the special cases later (Threads can be a field of other objects, which are inherited form GC object).

Reply:
Regarding the stack concept, the store needs to know the active Threads for marking. We could use some high level sets to insert/delete Threads, but this stack implementation is very fast in practice. When the stack cannot be used, the Threads can be marked by the store in other ways (I described an idea for that above).

Reply:
I think just using `std::set` of active threads should be fine.

Reply:
Ok, I have changed the patch to use std::set.